### PR TITLE
[DC-391] Person id validation cleaning rules

### PR DESCRIPTION
**Pull request by lrwb-aou**

### CODE REVIEW

File: data_steward/cleaners/query_generators/person_id_validator.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            clean_consts.DISPOSITION: bq_consts.WRITE_TRUNCATE,
+        }
+        query_list.append(query_dict)
+
+    return query_list
+
+if __name__ == '__main__':
+    import argparse
+
+    import cleaners.clean_cdr_engine
```
Comment (mark-velez) [inline data_steward/cleaners/query_generators/person_id_validator.py]: Codacy doesn't like this unused import

File: data_steward/cleaners/clean_cdr.py [hunk 2]
```diff
@@ -95,6 +100,12 @@ def _gather_unioned_ehr_queries(project, dataset):
 
 
 def clean_rdr_dataset(project=None, dataset=None):
+    """
+    Run all clean rules defined for the rdr dataset.
+
+    :param project:  Name of the BigQuery project.
```
[the remark this answers is not in the document]
Comment (lrwb-aou) [reply data_steward/cleaners/clean_cdr.py]: I'll make this change too.

File: data_steward/cleaners/clean_cdr.py [hunk 3]
```diff
@@ -151,8 +186,9 @@ def clean_ehr_rdr_de_identified_dataset(project=None, dataset=None):
 
 def get_dataset_and_project_names():
```
[the remark this answers is not in the document]
Comment (lrwb-aou) [reply data_steward/cleaners/clean_cdr.py]: Good idea.  We should plan this for future iterations.

File: data_steward/cleaners/query_generators/clean_years.py [hunk 4]
```diff
@@ -78,7 +78,7 @@ def get_year_of_birth_queries(project_id, dataset_id):
 
 if __name__ == '__main__':
     import argparse
-    import clean_cdr_engine as clean_engine
+    import cleaners.clean_cdr_engine as clean_engine
```
[the remark this answers is not in the document]
Comment (lrwb-aou) [reply data_steward/cleaners/query_generators/clean_years.py]: packages are renamed from `cleaners.query_generators` to `cdr_cleaner.cleaning_rules`

File: data_steward/cleaners/query_generators/person_id_validator.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    'measurement',
+    'procedure_occurrence',
+    'observation',
+    'device_exposure',
+    'specimen',
+]
+
+UNMAPPED_VALIDATION_TABLES = [
+    # mapping tables do not exist for the following tables
+    'death',
```
[the remark this answers is not in the document]
Comment (lrwb-aou) [reply data_steward/cleaners/query_generators/person_id_validator.py]: The person table is my reference table.  So I don't think it matters here.

File: data_steward/cleaners/query_generators/person_id_validator.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+]
+
+# The below query translates to:
```
[the remark this answers is not in the document]
Comment (lrwb-aou) [reply data_steward/cleaners/query_generators/person_id_validator.py]: Updated the comments and changed `NOT_CONSENTING_PERSON_IDS` to `EXISTING_AND_VALID_CONSENTING_RECORDS`.   Changed `DELETE_ORPHANED_PERSON_IDS` to `SELECT_EXISTING_PERSON_IDS`.

File: data_steward/cleaners/query_generators/person_id_validator.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    # get all unconsented non-ehr rows
+    'SELECT {fields} FROM `{project}.{dataset}.{table}` AS entry '
+    'RIGHT JOIN unconsented as cons on entry.person_id = cons.person_id '
+    'JOIN ppi_mappings AS maps ON maps.{table}_id = entry.{table}_id '
+)
+
+# drop rows of person_ids not in the person table
+DELETE_ORPHANED_PERSON_IDS = (
+    'SELECT {fields} FROM `{project}.{dataset}.{table}` AS entry '
+    'RIGHT JOIN `{project}.{dataset}.person` AS person '
```
[the remark this answers is not in the document]
Comment (lrwb-aou) [reply data_steward/cleaners/query_generators/person_id_validator.py]: `RIGHT JOIN` changed to `JOIN` in three places.  I'll get a jira issue created shortly.

**Comment by mark-velez** [comment]
@lrwb-aou thank you for adding the changes we discussed. Sorry, I'm only detecting this issue now- but I tested the query by running `get_person_id_validation_queries` and pasting one of the output queries in the BigQuery console. It complains that column name `person_id` is ambiguous:

image

Replacing it with `entry.person_id` seems to fix this. To be safe, you may want to qualify **all** output fields like this as there are other "foreign keys" which will present the same issue. For example, `condition_occurrence.provider_id`:

image

**Comment by lrwb-aou** [comment]
Thanks for catching that.  I had checked the update and delete queries.  I guess I forgot to check the select queries.

**Pull request merged**